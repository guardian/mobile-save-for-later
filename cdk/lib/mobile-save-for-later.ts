import { join } from "path";
import { GuApiGatewayWithLambdaByPath } from "@guardian/cdk";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import { GuLambdaFunction } from "@guardian/cdk/lib/constructs/lambda";
import type { App } from "aws-cdk-lib";
import { Duration } from "aws-cdk-lib";
import { CfnBasePathMapping, CfnDomainName } from "aws-cdk-lib/aws-apigateway";
import { PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { CfnRecordSetGroup } from "aws-cdk-lib/aws-route53";
import { CfnInclude } from "aws-cdk-lib/cloudformation-include";

export interface MobileSaveForLaterProps extends GuStackProps {
  certificateId: string;
  domainName: string;
  hostedZoneName: string;
  hostedZoneId: string;
  identityApiHost: string;
  reservedConcurrentExecutions: number;
}

export class MobileSaveForLater extends GuStack {
  constructor(scope: App, id: string, props: MobileSaveForLaterProps) {
    super(scope, id, props);

    const yamlTemplateFilePath = join(
      __dirname,
      "../..",
      "mobile-save-for-later/conf/cfn.yaml"
    );
    const yamlDefinedResources = new CfnInclude(this, "YamlTemplate", {
      templateFile: yamlTemplateFilePath,
    });

    const app = "mobile-save-for-later";

    const commonLambdaProps = {
      runtime: Runtime.JAVA_8, // We should upgrade to Java 11 in a future PR
      app,
      fileName: `${app}.jar`,
      // N.B. these lambdas previously used very specific, low memory sizes (384 & 394); odd choices for a Scala Lambda
    };

    const commonEnvironmentVariables = {
      App: app,
      Stack: this.stack,
      Stage: this.stage,
      IdentityApiHost: props.identityApiHost,
    };

    const saveArticlesLambda = new GuLambdaFunction(
      this,
      "save-articles-lambda",
      {
        handler: "com.gu.sfl.lambda.SaveArticlesLambda::handleRequest",
        functionName: `mobile-save-for-later-SAVE-cdk-${this.stage}`,
        timeout: Duration.seconds(60),
        environment: {
          ...commonEnvironmentVariables,
          SavedArticleLimit: "1000",
        },
        ...commonLambdaProps,
      }
    );

    const fetchArticlesLambda = new GuLambdaFunction(
      this,
      "fetch-articles-lambda",
      {
        handler: "com.gu.sfl.lambda.FetchArticlesLambda::handleRequest",
        functionName: `mobile-save-for-later-FETCH-cdk-${this.stage}`,
        timeout: Duration.seconds(20),
        reservedConcurrentExecutions: props.reservedConcurrentExecutions,
        environment: commonEnvironmentVariables,
        ...commonLambdaProps,
      }
    );

    [saveArticlesLambda, fetchArticlesLambda].map((lambda) => {
      lambda.addToRolePolicy(
        new PolicyStatement({
          actions: [
            "dynamodb:GetItem",
            "dynamodb:PutItem",
            "dynamodb:UpdateItem",
            "dynamodb:Query",
          ],
          resources: [
            yamlDefinedResources
              .getResource("SaveForLaterDynamoTable")
              // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-dynamodb-table.html#aws-resource-dynamodb-table-return-values
              .getAtt("Arn")
              .toString(),
          ],
        })
      );
    });

    const saveForLaterApi = new GuApiGatewayWithLambdaByPath(this, {
      app,
      restApiName: `${app}-api-${this.stage}`,
      monitoringConfiguration: { noMonitoring: true }, //TODO: configure monitoring (this is not setup in current CFN)
      targets: [
        {
          path: "/syncedPrefs/me/savedArticles",
          httpMethod: "POST",
          lambda: saveArticlesLambda,
        },
        {
          path: "/syncedPrefs/me",
          httpMethod: "GET",
          lambda: fetchArticlesLambda,
        },
      ],
    });

    // N.B. we cannot use GuCertificate here as we deploy to eu-west-1 but the certificate must be created in us-east-1.
    // https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-edge-optimized-custom-domain-name.html
    const certificateArn = `arn:aws:acm:us-east-1:${this.account}:certificate/${props.certificateId}`;

    const cfnDomainName = new CfnDomainName(this, "ApiDomainName", {
      domainName: props.domainName,
      certificateArn,
    });

    new CfnBasePathMapping(this, "ApiMapping", {
      domainName: cfnDomainName.ref,
      restApiId: saveForLaterApi.api.restApiId,
      stage: saveForLaterApi.api.deploymentStage.stageName,
    });

    new CfnRecordSetGroup(this, "ApiRoute53", {
      hostedZoneId: props.hostedZoneId,
      recordSets: [
        {
          name: props.domainName,
          type: "A",
          aliasTarget: {
            dnsName: cfnDomainName.attrDistributionDomainName,
            // This magical value is taken from the AWS docs:
            // https://docs.amazonaws.cn/en_us/AWSCloudFormation/latest/UserGuide/aws-properties-route53-aliastarget-1.html#aws-properties-route53-aliastarget-1-properties
            hostedZoneId: "Z2FDTNDATAQYW2",
          },
        },
      ],
    });

    // TODO:
    // Decide whether to port across or leave in CFN: Dynamo Table & Dynamo Throttle CloudWatch Alarms
  }
}
