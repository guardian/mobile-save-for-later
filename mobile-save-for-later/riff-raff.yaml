stacks: [mobile]
regions: [eu-west-1]

deployments:
  mobile-save-for-later:
    type: aws-lambda
    parameters:
      bucket: mobile-dist
      functionNames:
        - mobile-save-for-later-SAVE-
        - mobile-save-for-later-FETCH-
        - mobile-save-for-later-SAVE-cdk-
        - mobile-save-for-later-FETCH-cdk-
      fileName: mobile-save-for-later.jar
      prefixStack: false
    dependencies:
      - mobile-save-for-later-cfn
  mobile-save-for-later-cfn:
    type: cloud-formation
    app: mobile-save-for-later
    parameters:
      templateStagePaths:
        CODE: MobileSaveForLater-CODE.template.json
        PROD: MobileSaveForLater-PROD.template.json
