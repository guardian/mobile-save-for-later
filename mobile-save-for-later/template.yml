# This AWS SAM template has been generated from your function's configuration. If
# your function has one or more triggers, note that the AWS resources associated
# with these triggers aren't fully specified in this template and include
# placeholder values. Open this template in AWS Application Composer or your
# favorite IDE and modify it to specify a serverless application with other AWS
# resources.
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Description: An AWS Serverless Application Model template describing your function.
Resources:
  mobilesaveforlaterFETCHcdkCODE:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: 'target/scala-2.12/mobile-save-for-later.jar'
      Description: ''
      MemorySize: 1024
      Timeout: 20
      Handler: com.gu.sfl.lambda.FetchArticlesLambda::handleRequest
      Runtime: java11
      Architectures:
        - x86_64
      EphemeralStorage:
        Size: 512
      Environment:
        Variables:
          App: mobile-save-for-later
          APP: mobile-save-for-later
          IdentityApiHost: https://id.code.dev-guardianapis.com
          STACK: mobile
          Stack: mobile
          IdentityOktaAudience: https://profile.code.dev-theguardian.com/
          IdentityOktaIssuerUrl: https://profile.code.dev-theguardian.com/oauth2/aus3v9gla95Toj0EE0x7
          STAGE: CODE
          Stage: CODE
      EventInvokeConfig:
        MaximumEventAgeInSeconds: 21600
        MaximumRetryAttempts: 2
      PackageType: Zip
      Policies:
        - Statement:
            - Action:
                - s3:GetObject*
                - s3:GetBucket*
                - s3:List*
              Resource:
                - arn:aws:s3:::mobile-dist
                - arn:aws:s3:::mobile-dist/*
              Effect: Allow
            - Action:
                - ssm:GetParametersByPath
              Resource: >-
                arn:aws:ssm:eu-west-1:201359054765:parameter/CODE/mobile/mobile-save-for-later
              Effect: Allow
            - Action:
                - ssm:GetParameters
                - ssm:GetParameter
              Resource: >-
                arn:aws:ssm:eu-west-1:201359054765:parameter/CODE/mobile/mobile-save-for-later/*
              Effect: Allow
            - Action:
                - dynamodb:GetItem
                - dynamodb:PutItem
                - dynamodb:UpdateItem
                - dynamodb:Query
              Resource: >-
                arn:aws:dynamodb:eu-west-1:201359054765:table/mobile-save-for-later-CODE-articles
              Effect: Allow
            - Effect: Allow
              Action:
                - logs:CreateLogGroup
                - logs:CreateLogStream
                - logs:PutLogEvents
              Resource: '*'
          Architecture:
            x86_64
      ReservedConcurrentExecutions: 1
      SnapStart:
        ApplyOn: None
      Tags:
        App: mobile-save-for-later
        Stack: mobile
        Stage: CODE
        gu:build-tool: guardian/actions-riff-raff
        gu:cdk:version: 45.1.0
        gu:repo: guardian/mobile-save-for-later
      Events:
        Api1:
          Type: Api
          Properties:
            Path: /syncedPrefs/me
            Method: GET
        Api2:
          Type: Api
          Properties:
            Path: /syncedPrefs/me
            Method: GET

