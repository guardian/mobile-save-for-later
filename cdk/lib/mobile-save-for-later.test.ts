import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { MobileSaveForLater } from "./mobile-save-for-later";

describe("The MobileSaveForLater stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new MobileSaveForLater(app, "MobileSaveForLater", {
      stack: "mobile",
      stage: "TEST",
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
