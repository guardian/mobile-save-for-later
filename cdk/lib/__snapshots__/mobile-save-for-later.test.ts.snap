// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MobileSaveForLater stack matches the snapshot 1`] = `
Object {
  "Description": "Implements save for later for mobile",
  "Mappings": Object {
    "StageVariables": Object {
      "CODE": Object {
        "AlarmActionsEnabled": false,
        "ReservedConcurrency": 1,
        "TableReadCapacity": 1,
        "TableWriteCapacity": 1,
      },
      "PROD": Object {
        "AlarmActionsEnabled": true,
        "ReservedConcurrency": 50,
        "TableReadCapacity": 200,
        "TableWriteCapacity": 75,
      },
    },
  },
  "Parameters": Object {
    "App": Object {
      "Default": "mobile-save-for-later",
      "Description": "Application name",
      "Type": "String",
    },
    "CertArn": Object {
      "Description": "ACM Certificate ARN",
      "Type": "String",
    },
    "DeployBucket": Object {
      "Default": "mobile-dist",
      "Description": "S3 Bucket where riff-raff uploads artifacts on deploy",
      "Type": "String",
    },
    "DynamoNotificationTopic": Object {
      "Description": "SNS topic to notify when there's a dynamo throttling event",
      "Type": "String",
    },
    "HostedZoneId": Object {
      "Description": "Id of HostedZone",
      "Type": "String",
    },
    "HostedZoneName": Object {
      "Description": "Name of HostedZone",
      "Type": "String",
    },
    "IdentityApiHost": Object {
      "Description": "Identity App Host",
      "Type": "String",
    },
    "SavedArticleLimit": Object {
      "Description": "Saved article limit",
      "Type": "Number",
    },
    "Stack": Object {
      "Default": "mobile",
      "Description": "Stack name",
      "Type": "String",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ApiDomainName": Object {
      "Properties": Object {
        "CertificateArn": Object {
          "Ref": "CertArn",
        },
        "DomainName": Object {
          "Fn::Sub": "\${App}.\${HostedZoneName}",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/mobile-save-for-later",
          },
          Object {
            "Key": "Stack",
            "Value": "mobile",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "ApiMapping": Object {
      "Properties": Object {
        "DomainName": Object {
          "Ref": "ApiDomainName",
        },
        "RestApiId": Object {
          "Ref": "SaveForLaterApi",
        },
        "Stage": Object {
          "Ref": "Stage",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "ApiRoute53": Object {
      "Properties": Object {
        "HostedZoneId": Object {
          "Ref": "HostedZoneId",
        },
        "RecordSets": Array [
          Object {
            "AliasTarget": Object {
              "DNSName": Object {
                "Fn::GetAtt": Array [
                  "ApiDomainName",
                  "DistributionDomainName",
                ],
              },
              "HostedZoneId": "Z2FDTNDATAQYW2",
            },
            "Name": Object {
              "Ref": "ApiDomainName",
            },
            "Type": "A",
          },
        ],
      },
      "Type": "AWS::Route53::RecordSetGroup",
    },
    "FetchArticlesLambda": Object {
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DeployBucket",
          },
          "S3Key": Object {
            "Fn::Sub": "\${Stack}/\${Stage}/\${App}/\${App}.jar",
          },
        },
        "Environment": Object {
          "Variables": Object {
            "App": Object {
              "Ref": "App",
            },
            "IdentityApiHost": Object {
              "Ref": "IdentityApiHost",
            },
            "Stack": Object {
              "Ref": "Stack",
            },
            "Stage": Object {
              "Ref": "Stage",
            },
          },
        },
        "FunctionName": Object {
          "Fn::Sub": "\${App}-FETCH-\${Stage}",
        },
        "Handler": "com.gu.sfl.lambda.FetchArticlesLambda::handleRequest",
        "MemorySize": 394,
        "ReservedConcurrentExecutions": Object {
          "Fn::FindInMap": Array [
            "StageVariables",
            Object {
              "Ref": "Stage",
            },
            "ReservedConcurrency",
          ],
        },
        "Role": Object {
          "Fn::GetAtt": "SaveForLaterRole.Arn",
        },
        "Runtime": "java8",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/mobile-save-for-later",
          },
          Object {
            "Key": "lambda:createdBy",
            "Value": "SAM",
          },
          Object {
            "Key": "Stack",
            "Value": "mobile",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 20,
      },
      "Type": "AWS::Lambda::Function",
    },
    "FetchArticlesLambdaGetApiPermissionStage": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Ref": "FetchArticlesLambda",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Sub": Array [
            "arn:aws:execute-api:\${AWS::Region}:\${AWS::AccountId}:\${__ApiId__}/\${__Stage__}/GET/syncedPrefs/me",
            Object {
              "__ApiId__": Object {
                "Ref": "SaveForLaterApi",
              },
              "__Stage__": "*",
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "FetchArticlesLambdaPermission": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": "FetchArticlesLambda.Arn",
        },
        "Principal": "apigateway.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SaveArticlesLambda": Object {
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DeployBucket",
          },
          "S3Key": Object {
            "Fn::Sub": "\${Stack}/\${Stage}/\${App}/\${App}.jar",
          },
        },
        "Environment": Object {
          "Variables": Object {
            "App": Object {
              "Ref": "App",
            },
            "IdentityApiHost": Object {
              "Ref": "IdentityApiHost",
            },
            "SavedArticleLimit": Object {
              "Ref": "SavedArticleLimit",
            },
            "Stack": Object {
              "Ref": "Stack",
            },
            "Stage": Object {
              "Ref": "Stage",
            },
          },
        },
        "FunctionName": Object {
          "Fn::Sub": "\${App}-SAVE-\${Stage}",
        },
        "Handler": "com.gu.sfl.lambda.SaveArticlesLambda::handleRequest",
        "MemorySize": 384,
        "Role": Object {
          "Fn::GetAtt": "SaveForLaterRole.Arn",
        },
        "Runtime": "java8",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/mobile-save-for-later",
          },
          Object {
            "Key": "lambda:createdBy",
            "Value": "SAM",
          },
          Object {
            "Key": "Stack",
            "Value": "mobile",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SaveArticlesLambdaPostApiPermissionStage": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Ref": "SaveArticlesLambda",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Sub": Array [
            "arn:aws:execute-api:\${AWS::Region}:\${AWS::AccountId}:\${__ApiId__}/\${__Stage__}/POST/syncedPrefs/me/savedArticles",
            Object {
              "__ApiId__": Object {
                "Ref": "SaveForLaterApi",
              },
              "__Stage__": "*",
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SaveForLaterApi": Object {
      "Properties": Object {
        "Body": Object {
          "info": Object {
            "title": Object {
              "Fn::Sub": "\${App}-\${Stage}",
            },
            "version": "1.0.0",
          },
          "paths": Object {
            "/syncedPrefs/me": Object {
              "get": Object {
                "produces": Array [
                  "application/json",
                ],
                "responses": Object {
                  "200": Object {
                    "description": "200 response",
                  },
                },
                "x-amazon-apigateway-integration": Object {
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": Object {
                    "Fn::Sub": "arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${FetchArticlesLambda.Arn}/invocations",
                  },
                },
              },
            },
            "/syncedPrefs/me/savedArticles": Object {
              "post": Object {
                "consumes": Array [
                  "application/json",
                ],
                "produces": Array [
                  "application/json",
                ],
                "responses": Object {
                  "200": Object {
                    "description": "200 response",
                  },
                },
                "x-amazon-apigateway-integration": Object {
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": Object {
                    "Fn::Sub": "arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${SaveArticlesLambda.Arn}/invocations",
                  },
                },
              },
            },
          },
          "swagger": "2.0",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/mobile-save-for-later",
          },
          Object {
            "Key": "Stack",
            "Value": "mobile",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "SaveForLaterApiDeployment7818545d6f": Object {
      "Properties": Object {
        "Description": "RestApi deployment id: 7818545d6f935a518be3b1f7ecebec1729443529",
        "RestApiId": Object {
          "Ref": "SaveForLaterApi",
        },
        "StageName": "Stage",
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "SaveForLaterApiStage": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "SaveForLaterApiDeployment7818545d6f",
        },
        "RestApiId": Object {
          "Ref": "SaveForLaterApi",
        },
        "StageName": Object {
          "Ref": "Stage",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/mobile-save-for-later",
          },
          Object {
            "Key": "Stack",
            "Value": "mobile",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "SaveForLaterDynamoTable": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "userId",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "userId",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": Object {
            "Fn::FindInMap": Array [
              "StageVariables",
              Object {
                "Ref": "Stage",
              },
              "TableReadCapacity",
            ],
          },
          "WriteCapacityUnits": Object {
            "Fn::FindInMap": Array [
              "StageVariables",
              Object {
                "Ref": "Stage",
              },
              "TableWriteCapacity",
            ],
          },
        },
        "TableName": Object {
          "Fn::Sub": "\${App}-\${Stage}-articles",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/mobile-save-for-later",
          },
          Object {
            "Key": "Stack",
            "Value": "mobile",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
    },
    "SaveForLaterReadThrottleEvents": Object {
      "Properties": Object {
        "ActionsEnabled": Object {
          "Fn::FindInMap": Array [
            "StageVariables",
            Object {
              "Ref": "Stage",
            },
            "AlarmActionsEnabled",
          ],
        },
        "AlarmActions": Array [
          Object {
            "Ref": "DynamoNotificationTopic",
          },
        ],
        "AlarmName": Object {
          "Fn::Sub": "\${App}-\${Stage}-articles-ReadCapacityUnitsLimit-BasicAlarm",
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "TableName",
            "Value": Object {
              "Ref": "SaveForLaterDynamoTable",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ReadThrottleEvents",
        "Namespace": "AWS/DynamoDB",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "SaveForLaterRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
        },
        "Path": "/",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Object {
                "Action": Array [
                  "logs:CreateLogGroup",
                  "logs:CreateLogStream",
                  "logs:PutLogEvents",
                ],
                "Effect": "Allow",
                "Resource": "*",
              },
            },
            "PolicyName": "logs",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Object {
                "Action": Array [
                  "ssm:GetParametersByPath",
                ],
                "Effect": "Allow",
                "Resource": Object {
                  "Fn::Sub": "arn:aws:ssm:\${AWS::Region}:\${AWS::AccountId}:parameter/\${App}/\${Stage}/\${Stack}",
                },
              },
            },
            "PolicyName": "ssm-config",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Object {
                "Action": Array [
                  "dynamodb:GetItem",
                  "dynamodb:PutItem",
                  "dynamodb:UpdateItem",
                  "dynamodb:Query",
                ],
                "Effect": "Allow",
                "Resource": Object {
                  "Fn::Sub": "arn:aws:dynamodb:\${AWS::Region}:\${AWS::AccountId}:table/\${App}-\${Stage}-articles",
                },
              },
            },
            "PolicyName": "dynamo-action",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Object {
                "Action": Array [
                  "cloudwatch:PutMetricData",
                ],
                "Effect": "Allow",
                "Resource": "*",
              },
            },
            "PolicyName": "metrics",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/mobile-save-for-later",
          },
          Object {
            "Key": "Stack",
            "Value": "mobile",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SaveForLaterWriteThrottleEvents": Object {
      "Properties": Object {
        "ActionsEnabled": Object {
          "Fn::FindInMap": Array [
            "StageVariables",
            Object {
              "Ref": "Stage",
            },
            "AlarmActionsEnabled",
          ],
        },
        "AlarmActions": Array [
          Object {
            "Ref": "DynamoNotificationTopic",
          },
        ],
        "AlarmName": Object {
          "Fn::Sub": "\${App}-\${Stage}-articles-WriteCapacityUnitsLimit-BasicAlarm",
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "TableName",
            "Value": Object {
              "Ref": "SaveForLaterDynamoTable",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "WriteThrottleEvents",
        "Namespace": "AWS/DynamoDB",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;
